### PR TITLE
Lint/AssignmentInCondition

### PR DESCRIPTION
**Pull request by mhashizume**
This commit takes variable assignments in conditionals and wraps them in parentheses to make the intention more explicit.

### CODE REVIEW

File: lib/puppet/pops/loader/static_loader.rb [hunk 1]
```diff
@@ -46,7 +46,7 @@ def initialize
 
   def discover(type, error_collector = nil, name_authority = Pcore::RUNTIME_NAME_AUTHORITY)
     # Static loader only contains runtime types
-    return EMPTY_ARRAY unless type == :type && name_authority == name_authority = Pcore::RUNTIME_NAME_AUTHORITY
+    return EMPTY_ARRAY unless type == :type && name_authority == (name_authority = Pcore::RUNTIME_NAME_AUTHORITY)
```
[the remark this answers is not in the document]
Comment (mhashizume) [reply lib/puppet/pops/loader/static_loader.rb:49]: Yeah this was kind of fuzzy for me too. Maybe if #discover is called and the positional argument is overridden, this exists as a secondary check? I'm not sure.

[the remark this answers is not in the document]
Comment (mhashizume) [reply lib/puppet/pops/loader/static_loader.rb:49]: I wasn't really able to tell the intent from the original commit: https://github.com/puppetlabs/puppet/commit/89c291e7a7a037d92c0976a84806acdbcc618f24

But FWIW we have just disabled this check on this code in the past as well: https://github.com/puppetlabs/puppet/commit/6c257fc7827989c2af2901f974666f0f23611153

[the remark this answers is not in the document]
Comment (joshcooper) [reply lib/puppet/pops/loader/static_loader.rb:49]: I naively assumed the assignment would occur first and then both `name_authority` variables would refer to the same value, but it doesn't work that way:

```
irb(main):006:0> a = 1
=> 1
irb(main):007:0> a == (a = 2)
=> false
```

I think this is a bug in the original code...  Note how this assignment:

```ruby
return EMPTY_ARRAY unless type == :type && name_authority == name_authority = Pcore::RUNTIME_NAME_AUTHORITY
```

Affects the equality check in the next expression:

```ruby
typed_names = type == :type && name_authority == Pcore::RUNTIME_NAME_AUTHORITY ? @loaded.keys : EMPTY_ARRAY
```

IOW  `name_authority == Pcore::RUNTIME_NAME_AUTHORITY` is always true. I think the intended code was supposed to be:

```ruby
return EMPTY_ARRAY unless type == :type && name_authority == Pcore::RUNTIME_NAME_AUTHORITY
```

Which means the following conditional is redundant:

```ruby
typed_names = type == :type && name_authority == Pcore::RUNTIME_NAME_AUTHORITY ? @loaded.keys : EMPTY_ARRAY
```

I think the method is supposed to be:

```ruby
    return EMPTY_ARRAY unless type == :type && name_authority == Pcore::RUNTIME_NAME_AUTHORITY

    typed_names = @loaded.keys 
    block_given? ? typed_names.select { |tn| yield(tn) } : typed_names
```

[the remark this answers is not in the document]
Comment (mhashizume) [reply lib/puppet/pops/loader/static_loader.rb]: Thanks for looking into that Josh, I was having a hard time deciphering what was going on.

**Review by joshcooper — APPROVED** [review]

**Pull request merged**